### PR TITLE
feat(grouping): Add `hashing_metadata` field to `GroupHashMetadata` table

### PR DESCRIPTION
**Pull request by lobsterkatie**
This adds a new field, `hashing_metadata`, to the `GroupHashMetadata` table, to serve as a compliment to the `hash_basis` field added in https://github.com/getsentry/sentry/pull/79835. 

Whereas that field stores the overall grouping method (stacktrace, message, custom fingerprint, etc.), this new field will store more detailed, grouping-method specific data. For example, when grouping on message, it will store whether or not the message was parameterized; when grouping on fingerprint, it will store the source of the fingerprint; when grouping on stacktrace, it will note whether the stacktrace was found in an exception or in a thread. (The full scope of the data stored for each grouping method can be found in the `XYZHashingMetadata` types added in this PR.)

Code to add data to this field is included in https://github.com/getsentry/sentry/pull/80534.

### CODE REVIEW

**Review by wedamija — APPROVED** [review]

**Comment by asottile-sentry** [comment]
when using a stacked workflow please keep the PR permanently in draft state or it will ping the entire company

File: src/sentry/types/grouphash_metadata.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+# NOTE: The structure in these metadata types is intentionaly flat, to make it easier to query in
+# Redash or BigQuery, and they are all merged into a flat type, stored in
+# `GroupHashMetadata.hashing_metadata`. If entries are added, they should be namespaced according to
+# their corresponding hash basis (so, for example, `fingerprint_source` and `message_source`, rather
+# than just `source`), both for clarity and to avoid collisions.
+
+
+# Fingerprint data, gathered both during stand-alone custom/built-in fingerprinting and hybrid
+# fingerprinting involving message, stacktrace, security, or template hashing.
+class FingerprintHashingMetadata(TypedDict):
```
Comment (JoshFerge) [inline src/sentry/types/grouphash_metadata.py:14]: these comments may be easier to read as docstrings (wrapped in `"""`s)

**Review by JoshFerge — CHANGES_REQUESTED** [review]
if we use GZipDictField, these parameters won't be queryable in redash/bigquery, which means that overall our idea to use postgres as a datastore won't work.

if we're concerned about storage size, we should consider a different solution, otherwise i think we need to make this a field that isn't gzipped.

**Comment by lobsterkatie** [comment]
> if we use GZipDictField, these parameters won't be queryable in redash/bigquery, which means that overall our idea to use postgres as a datastore won't work.

Right! _That's_ why I originally had it as a JSON field. I couldn't remember why I'd chosen that over the gzipdict, and so I swapped it in order to get the defaults-to-an-empty-dict-rather-than-null behavior. But past-past-me was smarter than recent-past me - you're absolutely right, it should be JSON.

Thank you for catching that!

File: src/sentry/models/grouphashmetadata.py [hunk 2]
```diff
@@ -56,6 +58,12 @@ class GroupHashMetadata(Model):
     latest_grouping_config = models.CharField(null=True)
     # The primary grouping method (message, stacktrace, fingerprint, etc.)
     hash_basis = models.CharField(choices=HashBasis, null=True)
+    # Metadata about the inputs to the hashing process and the hashing process itself (what
+    # fingerprinting rules were matched? did we parameterize the message? etc.). For the specific
+    # data stored, see the class definitions of the `HashingMetadata` subtypes.
+    hashing_metadata: models.Field[HashingMetadata | None, HashingMetadata | None] = JSONField(
```
Comment (JoshFerge) [inline src/sentry/models/grouphashmetadata.py:64]: why does this need to be repeated `HashingMetadata | None, HashingMetadata | None`?

**Review by JoshFerge — APPROVED** [review]

File: src/sentry/models/grouphashmetadata.py (hunk 2 above)
Comment (lobsterkatie) [reply src/sentry/models/grouphashmetadata.py:64]: Not well documented, but this is the setter type and the getter type: https://github.com/typeddjango/django-stubs/blob/ab4dcfa002dbe138b7548086328f13cbde3c41b3/django-stubs/db/models/fields/__init__.pyi#L46-L51.

**Pull request merged**